pkg/cmd/provisioning/identity_provider.go

```diff
@@ -58,145 +64,244 @@ var (
+	if generateOnly {
+		oidcIdentityProviderJSON := fmt.Sprintf(iamIdentityProviderTemplate, issuerURL)
 
-	if err != nil {
-		return errors.Wrapf(err, "Failed to upload JSON web key set (JWKS) in the S3 bucket %s", bucketName)
-	}
-	log.Printf("JSON web key set (JWKS) in the S3 bucket %s at %s updated", bucketName, keysURI)
+		err := saveToFile("AWS IAM Identity Provider", filepath.Join(targetDir, iamIdentityProviderFilename), []byte(oidcIdentityProviderJSON))
+		if err != nil {
+			errors.Wrap(err, "failed to save IAM Identity Provider file")
+		}
 
-	err = saveToFile("JSON web key set (JWKS)", filepath.Join(targetDir, "keys.json"), jwks)
-	if err != nil {
-		return err
+	} else {
+		oidcProviderList, err := client.ListOpenIDConnectProviders(&iam.ListOpenIDConnectProvidersInput{})
+		if err != nil {
+			return errors.Wrap(err, "Failed to fetch list of Identity Providers")
+		}
+
+		var providerARN string
+		for _, provider := range oidcProviderList.OpenIDConnectProviderList {
+			ok, err := isExistingIdentifyProvider(client, *provider.Arn, namePrefix)
+			if err != nil {
+				return errors.Wrapf(err, "Failed to check existing Identity Provider %s", *provider.Arn)
+			}
+
+			if ok {
+				providerARN = *provider.Arn
+				log.Printf("Existing Identity Provider found with ARN: %s", providerARN)
+				break
+			}
+		}
+
+		if len(providerARN) == 0 {
+			oidcOutput, err := client.CreateOpenIDConnectProvider(&iam.CreateOpenIDConnectProviderInput{
+				ClientIDList: []*string{
+					awssdk.String("openshift"),
+					awssdk.String("sts.amazonaws.com"),
+				},
+				ThumbprintList: []*string{
+					awssdk.String("A9D53002E97E00E043244F3D170D6F4C414104FD"), // root CA thumbprint for Amazon S3 (DigiCert)
+				},
+				Url: awssdk.String(issuerURL),
+			})
+			if err != nil {
+				return errors.Wrap(err, "Failed to create Identity Provider")
+			}
+
+			providerARN = *oidcOutput.OpenIDConnectProviderArn
+
+			_, err = client.TagOpenIDConnectProvider(&iam.TagOpenIDConnectProviderInput{
+				OpenIDConnectProviderArn: &providerARN,
+				Tags: []*iam.Tag{
+					{
+						Key:   awssdk.String(fmt.Sprintf("%s/%s", ccoctlAWSResourceTagKeyPrefix, namePrefix)),
+						Value: awssdk.String(ownedCcoctlAWSResourceTagValue),
+					},
+				},
+			})
+			if err != nil {
+				return errors.Wrapf(err, "Failed to tag the identity provider with arn: %s", providerARN)
+			}
+
+			log.Printf("Identity Provider created with ARN: %s", providerARN)
+		}
 	}
+	return nil
+}
 
-	oidcProviderList, err := client.ListOpenIDConnectProviders(&iam.ListOpenIDConnectProvidersInput{})
+func createJSONWebKeySet(client aws.Client, publicKeyFilepath, bucketName, namePrefix, targetDir string, generateOnly bool) error {
+	jwks, err := buildJsonWebKeySet(publicKeyFilepath)
 	if err != nil {
-		return errors.Wrap(err, "Failed to fetch list of Identity Providers")
+		return errors.Wrap(err, "Failed to build JSON web key set from the public key")
```

Comment:
https://github.com/golang/go/wiki/CodeReviewComments#error-strings error wraps should not start with capital letters.